tinygrad/codegen/optimizer.py

```diff
@@ -166,119 +166,33 @@ def hand_coded_optimizations(self):
     self.simplify_ones()
 
-    # should use HIP tensor cores?
-    if getenv("TC", 1) != 0 and self.bufs[0].device == "HIP" and self.reduceop and self.reduceop.op == ReduceOps.SUM and \
-        isinstance(self.reduceop.src[0], LazyOp) and self.reduceop.src[0].op == UnaryOps.CAST and \
-        isinstance(self.reduceop.src[0].src[0], LazyOp) and self.reduceop.src[0].src[0].op == BinaryOps.MUL and \
-        isinstance(self.reduceop.src[0].src[0].src[0], LazyBuffer) and isinstance(self.reduceop.src[0].src[0].src[1], LazyBuffer) and self.opts.has_local and \
-        self.reduceop.src[0].src[0].src[0].dtype == dtypes.half and self.reduceop.src[0].src[0].src[1].dtype == dtypes.half:
-      # HIP tensor cores are 16x16x16
-      buf0 = self.bufs.index(self.reduceop.src[0].src[0].src[0])
-      buf1 = self.bufs.index(self.reduceop.src[0].src[0].src[1])
-      buf0_strides = self.sts[buf0].real_strides()
-      buf1_strides = self.sts[buf1].real_strides()
-      axis_buf0 = [(i,self.full_shape[i],buf1_strides[i]) for i,s in enumerate(buf0_strides) if s == 0 and self.full_shape[i]%16 == 0 and i < self.first_reduce]
-      axis_buf1 = [(i,self.full_shape[i],buf0_strides[i]) for i,s in enumerate(buf1_strides) if s == 0 and self.full_shape[i]%16 == 0 and i < self.first_reduce]
-      if axis_buf0 and axis_buf1 and self.full_shape[self.first_reduce]%8 == 0 and (self.shape_len-self.first_reduce) == 1:
-        if DEBUG >= 3: print("HIP TENSOR CORES", axis_buf0, axis_buf1)
-        self.use_tensor_cores = getenv("TC", 1) == 1  # TC=2 will do the shape ops without the WMMA
-        self.reverse_upcast_dir = True
-
-        # TODO: select axis in smart way
-        s0, s1 = axis_buf0[-1][0], axis_buf1[-1][0]
+    if getenv("TC", 1) != 0 and self.opts.has_local and self.reduceop and self.reduceop.op == ReduceOps.SUM and self.bufs[0].device in tensor_cores:
+      for tc in tensor_cores[self.bufs[0].device]:
+        if not((tc.arch is None or tc.arch == os.uname().machine) and isinstance(self.reduceop.src[0], LazyOp)):
+          continue
+        has_cast = tc.dtype_in != tc.dtype_out
+        if has_cast and not(isinstance(self.reduceop.src[0], LazyOp) and self.reduceop.src[0].op == UnaryOps.CAST and self.reduceop.src[0].arg[0] == tc.dtype_out):
+          continue
+        mul_op = self.reduceop.src[0].src[0] if has_cast else self.reduceop.src[0]
+        if not(isinstance(mul_op, LazyOp) and mul_op.op == BinaryOps.MUL and isinstance(mul_op.src[0], LazyBuffer) and mul_op.src[0].dtype == tc.dtype_in and isinstance(mul_op.src[1], LazyBuffer) and mul_op.src[1].dtype == tc.dtype_in):
+          continue
+        buf0, buf1 = self.bufs.index(mul_op.src[0]), self.bufs.index(mul_op.src[1])
+        buf0_strides, buf1_strides = self.sts[buf0].real_strides(), self.sts[buf1].real_strides()
+        axis_buf0 = [(i,self.full_shape[i],buf1_strides[i]) for i,s in enumerate(buf0_strides) if s == 0 and self.full_shape[i]%tc.dims[0] == 0 and i < self.first_reduce]
```

Comment:
You can speed this up by only enumerating until `i == self.first_reduce` instead of checking all. 
`enumerate(buf0_strides[:self.first_reduce])`

Reply:
Done!